internal/headertemplate/headertemplate.go

```diff
@@ -0,0 +1,239 @@
+// Package headertemplate contains functions for rendering header templates.
+package headertemplate
+
+import "strings"
+
+// Render renders a header template string.
+func Render(src string, fn func(ref []string) string) string {
+	p := newParser(src, fn)
+	return p.parse()
+}
+
+type parser struct {
```

Comment:
could you add a comment here with a reference ebnf grammar?

Reply:
I don't have one. Do you want me to write one?

Reply:
Yes that would be helpful. Just an equivalent grammar for reference in a comment somewhere here

Reply:
I added a peg grammar.